packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart

```diff
@@ -0,0 +1,418 @@
+
+import '../../globals.dart';
+import '../../service_extensions.dart';
+import '../../service_registrations.dart';
+import '../../utils.dart';
+import 'flutter_icon_renderer.dart';
+
+/// Group of buttons where each button toggles the state of a VMService
+/// extension.
+///
+/// Use this class any time you need to write UI code that exposes button(s) to
+/// control VMServiceExtension. This class handles error handling and lifecycle
+/// states for keeping state on the client and device consistent so you don't
+/// have to.
+class ServiceExtensionButtonGroup extends StatefulWidget {
+  const ServiceExtensionButtonGroup({
+    this.minIncludeLabelWidth,
+    @required this.extensions,
+  });
+
+  final double minIncludeLabelWidth;
+  final List<ToggleableServiceExtensionDescription> extensions;
+
+  @override
+  _ServiceExtensionButtonGroupState createState() =>
+      _ServiceExtensionButtonGroupState();
+}
+
+/// Data class tracking the state of a single service extension.
+class ExtensionState {
+  ExtensionState(this.description);
+
+  final ToggleableServiceExtensionDescription description;
+  bool isSelected = false;
+  bool isAvailable = false;
+  StreamSubscription subscription;
+
+  void dispose() {
+    subscription?.cancel();
+    subscription = null;
+  }
+}
+
+class _ServiceExtensionButtonGroupState
+    extends State<ServiceExtensionButtonGroup> {
+  List<ExtensionState> _extensionStates;
+
+  @override
+  void initState() {
+    super.initState();
+    // To use ToggleButtons we have to track states for all buttons in the
+    // group here rather than tracking state with the individual button widgets
+    // which would be more natural.
+
+    _extensionStates = [for (var e in widget.extensions) ExtensionState(e)];
+
+    for (var extension in _extensionStates) {
+      // Listen for changes to the state of each service extension using the
+      // VMServiceManager.
+      final extensionName = extension.description.extension;
+      // Update the button state to match the latest state on the VM.
+      serviceManager.serviceExtensionManager
+          .getServiceExtensionState(extensionName, (state) {
+        setState(() {
+          extension.isSelected =
+              state.value == extension.description.enabledValue;
+        });
+      });
+      // Track whether the extension is actually exposed by the VM.
+      extension.isAvailable = serviceManager.serviceExtensionManager
+          .isServiceExtensionAvailable(extensionName);
+      extension.subscription =
+          serviceManager.serviceExtensionManager.hasServiceExtension(
+        extensionName,
+        (available) {
+          setState(() {
+            extension.isAvailable = available;
+          });
+        },
+      );
+    }
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+    for (var config in _extensionStates) {
+      config.dispose();
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
```

Comment:
Thinking about this one, it seems a bit odd that one class is deciding both how to talk to the service extensions as well as what the service extension buttons should look like.  For example, we have the hot reload buttons, which look very different from the inspector buttons.

This might be a good case to use a `mixin on State` pattern.  `TickerProviderStateMixin` follows this pattern in Flutter, and it might be useful here too.

Then, when we add in buttons that talk to the service extension, we could talk to the service extension simply:

```
class IconServiceButton extends StatefulWidget {
  ISB(this.icon, this.extension);
}

class ISBState extends State<ISB> with ServiceExtensionMixin {
  void initState() {
    super.initState();
    registerExtension(widget.extension);
  }

  // [toggleExtension] is a protected method on SEM.
  Widget build(context) => IconButton(icon, onPressed: toggleExtension);
}
```

Different buttons can all be defined with different appearances and the same backing code for talking to the service extension.